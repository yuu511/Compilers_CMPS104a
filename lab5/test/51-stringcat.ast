ROOT "<<ROOT>>" 0.0.0
| FUNCTION "" 6.22.1
| | TYPE_ID "" 6.22.1
| | | VOID "void" 6.22.1
| | | IDENT "fail" 6.22.6
| | PARAM "(" 6.22.11
| | | TYPE_ID "" 6.22.12
| | | | STRING "string" 6.22.12
| | | | IDENT "expr" 6.22.19
| | | TYPE_ID "" 6.22.25
| | | | STRING "string" 6.22.25
| | | | IDENT "file" 6.22.32
| | | TYPE_ID "" 6.22.38
| | | | INT "int" 6.22.38
| | | | IDENT "line" 6.22.42
| FUNCTION "" 6.24.1
| | TYPE_ID "" 6.24.1
| | | VOID "void" 6.24.1
| | | IDENT "putchr" 6.24.6
| | PARAM "(" 6.24.13
| | | TYPE_ID "" 6.24.14
| | | | INT "int" 6.24.14
| | | | IDENT "chr" 6.24.18
| FUNCTION "" 6.25.1
| | TYPE_ID "" 6.25.1
| | | VOID "void" 6.25.1
| | | IDENT "putint" 6.25.6
| | PARAM "(" 6.25.13
| | | TYPE_ID "" 6.25.14
| | | | INT "int" 6.25.14
| | | | IDENT "num" 6.25.18
| FUNCTION "" 6.26.1
| | TYPE_ID "" 6.26.1
| | | VOID "void" 6.26.1
| | | IDENT "putstr" 6.26.6
| | PARAM "(" 6.26.13
| | | TYPE_ID "" 6.26.14
| | | | STRING "string" 6.26.14
| | | | IDENT "str" 6.26.21
| FUNCTION "" 6.28.1
| | TYPE_ID "" 6.28.1
| | | INT "int" 6.28.1
| | | IDENT "getchr" 6.28.5
| | PARAM "(" 6.28.11
| FUNCTION "" 6.29.1
| | TYPE_ID "" 6.29.1
| | | STRING "string" 6.29.1
| | | IDENT "getstr" 6.29.8
| | PARAM "(" 6.29.14
| FUNCTION "" 6.30.1
| | TYPE_ID "" 6.30.1
| | | STRING "string" 6.30.1
| | | IDENT "getln" 6.30.8
| | PARAM "(" 6.30.13
| FUNCTION "" 7.8.1
| | TYPE_ID "" 7.8.1
| | | INT "int" 7.8.1
| | | IDENT "strlen" 7.8.5
| | PARAM "(" 7.8.12
| | | TYPE_ID "" 7.8.13
| | | | STRING "string" 7.8.13
| | | | IDENT "str" 7.8.20
| | BLOCK "{" 7.8.25
| | | TYPE_ID "" 7.9.4
| | | | INT "int" 7.9.4
| | | | IDENT "len" 7.9.8
| | | | INTCON "0" 7.9.14
| | | WHILE "while" 7.10.4
| | | | NE "!=" 7.10.20
| | | | | INDEX "[" 7.10.14
| | | | | | IDENT "str" 7.10.11
| | | | | | IDENT "len" 7.10.15
| | | | | CHARCON "'\0'" 7.10.23
| | | | '=' "=" 7.10.33
| | | | | IDENT "len" 7.10.29
| | | | | '+' "+" 7.10.39
| | | | | | IDENT "len" 7.10.35
| | | | | | INTCON "1" 7.10.41
| | | RETURN "return" 7.11.4
| | | | IDENT "len" 7.11.11
| FUNCTION "" 7.14.1
| | TYPE_ID "" 7.14.1
| | | VOID "void" 7.14.1
| | | IDENT "strcat" 7.14.6
| | PARAM "(" 7.14.13
| | | TYPE_ID "" 7.14.14
| | | | STRING "string" 7.14.14
| | | | IDENT "dest" 7.14.21
| | | TYPE_ID "" 7.14.27
| | | | STRING "string" 7.14.27
| | | | IDENT "src" 7.14.34
| | BLOCK "{" 7.14.39
| | | TYPE_ID "" 7.15.4
| | | | INT "int" 7.15.4
| | | | IDENT "pos" 7.15.8
| | | | CALL "(" 7.15.21
| | | | | IDENT "strlen" 7.15.14
| | | | | IDENT "dest" 7.15.22
| | | TYPE_ID "" 7.16.4
| | | | INT "int" 7.16.4
| | | | IDENT "srcix" 7.16.8
| | | | INTCON "0" 7.16.16
| | | WHILE "while" 7.17.4
| | | | NE "!=" 7.17.22
| | | | | INDEX "[" 7.17.14
| | | | | | IDENT "src" 7.17.11
| | | | | | IDENT "srcix" 7.17.15
| | | | | CHARCON "'\0'" 7.17.25
| | | | BLOCK "{" 7.17.31
| | | | | '=' "=" 7.18.17
| | | | | | INDEX "[" 7.18.11
| | | | | | | IDENT "dest" 7.18.7
| | | | | | | IDENT "pos" 7.18.12
| | | | | | INDEX "[" 7.18.22
| | | | | | | IDENT "src" 7.18.19
| | | | | | | IDENT "srcix" 7.18.23
| | | | | '=' "=" 7.19.11
| | | | | | IDENT "pos" 7.19.7
| | | | | | '+' "+" 7.19.17
| | | | | | | IDENT "pos" 7.19.13
| | | | | | | INTCON "1" 7.19.19
| | | | | '=' "=" 7.20.13
| | | | | | IDENT "srcix" 7.20.7
| | | | | | '+' "+" 7.20.21
| | | | | | | IDENT "srcix" 7.20.15
| | | | | | | INTCON "1" 7.20.23
| | | '=' "=" 7.22.14
| | | | INDEX "[" 7.22.8
| | | | | IDENT "dest" 7.22.4
| | | | | IDENT "pos" 7.22.9
| | | | CHARCON "'\0'" 7.22.16
| FUNCTION "" 7.25.1
| | TYPE_ID "" 7.25.1
| | | INT "int" 7.25.1
| | | IDENT "main" 7.25.5
| | PARAM "(" 7.25.10
| | | TYPE_ID "" 7.25.11
| | | | INT "int" 7.25.11
| | | | IDENT "argc" 7.25.15
| | | TYPE_ID "" 7.25.21
| | | | ARRAY "array" 7.25.21
| | | | | STRING "string" 7.25.27
| | | | IDENT "argv" 7.25.35
| | BLOCK "{" 7.25.41
| | | TYPE_ID "" 7.26.4
| | | | INT "int" 7.26.4
| | | | IDENT "length" 7.26.8
| | | | INTCON "0" 7.26.17
| | | TYPE_ID "" 7.27.4
| | | | INT "int" 7.27.4
| | | | IDENT "argi" 7.27.8
| | | | INTCON "1" 7.27.15
| | | WHILE "while" 7.28.4
| | | | LT "<" 7.28.16
| | | | | IDENT "argi" 7.28.11
| | | | | IDENT "argc" 7.28.18
| | | | BLOCK "{" 7.28.24
| | | | | '=' "=" 7.29.14
| | | | | | IDENT "length" 7.29.7
| | | | | | '+' "+" 7.29.45
| | | | | | | '+' "+" 7.29.23
| | | | | | | | IDENT "length" 7.29.16
| | | | | | | | CALL "(" 7.29.32
| | | | | | | | | IDENT "strlen" 7.29.25
| | | | | | | | | INDEX "[" 7.29.37
| | | | | | | | | | IDENT "argv" 7.29.33
| | | | | | | | | | IDENT "argi" 7.29.38
| | | | | | | INTCON "2" 7.29.47
| | | | | '=' "=" 7.30.12
| | | | | | IDENT "argi" 7.30.7
| | | | | | '+' "+" 7.30.19
| | | | | | | IDENT "argi" 7.30.14
| | | | | | | INTCON "1" 7.30.21
| | | CALL "(" 7.32.11
| | | | IDENT "putstr" 7.32.4
| | | | STRINGCON ""concat length = "" 7.32.12
| | | CALL "(" 7.33.11
| | | | IDENT "putint" 7.33.4
| | | | IDENT "length" 7.33.12
| | | CALL "(" 7.34.11
| | | | IDENT "putchr" 7.34.4
| | | | CHARCON "'\n'" 7.34.12
| | | TYPE_ID "" 7.35.4
| | | | STRING "string" 7.35.4
| | | | IDENT "concat" 7.35.11
| | | | ALLOC "alloc" 7.35.20
| | | | | STRING "string" 7.35.26
| | | | | IDENT "length" 7.35.35
| | | '=' "=" 7.36.9
| | | | IDENT "argi" 7.36.4
| | | | INTCON "1" 7.36.11
| | | CALL "(" 7.37.11
| | | | IDENT "putchr" 7.37.4
| | | | CHARCON "'['" 7.37.12
| | | WHILE "while" 7.38.4
| | | | LT "<" 7.38.16
| | | | | IDENT "argi" 7.38.11
| | | | | IDENT "argc" 7.38.18
| | | | BLOCK "{" 7.38.24
| | | | | CALL "(" 7.39.14
| | | | | | IDENT "strcat" 7.39.7
| | | | | | IDENT "concat" 7.39.15
| | | | | | STRINGCON ""("" 7.39.23
| | | | | CALL "(" 7.40.14
| | | | | | IDENT "strcat" 7.40.7
| | | | | | IDENT "concat" 7.40.15
| | | | | | INDEX "[" 7.40.27
| | | | | | | IDENT "argv" 7.40.23
| | | | | | | IDENT "argi" 7.40.28
| | | | | CALL "(" 7.41.14
| | | | | | IDENT "strcat" 7.41.7
| | | | | | IDENT "concat" 7.41.15
| | | | | | STRINGCON "")"" 7.41.23
| | | | | '=' "=" 7.42.12
| | | | | | IDENT "argi" 7.42.7
| | | | | | '+' "+" 7.42.19
| | | | | | | IDENT "argi" 7.42.14
| | | | | | | INTCON "1" 7.42.21
| | | CALL "(" 7.44.11
| | | | IDENT "putstr" 7.44.4
| | | | IDENT "concat" 7.44.12
| | | CALL "(" 7.45.11
| | | | IDENT "putstr" 7.45.4
| | | | STRINGCON ""]\n"" 7.45.12
| | | RETURN "return" 7.46.4
| | | | INTCON "0" 7.46.11
