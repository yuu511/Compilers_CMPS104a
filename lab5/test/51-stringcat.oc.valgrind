==6656== Memcheck, a memory error detector
==6656== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==6656== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==6656== Command: ../oc 51-stringcat.oc
==6656== 
==6656== 
==6656== HEAP SUMMARY:
==6656==     in use at exit: 0 bytes in 0 blocks
==6656==   total heap usage: 1,180 allocs, 1,180 frees, 189,458 bytes allocated
==6656== 
==6656== All heap blocks were freed -- no leaks are possible
==6656== 
==6656== For counts of detected and suppressed errors, rerun with: -v
==6656== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
