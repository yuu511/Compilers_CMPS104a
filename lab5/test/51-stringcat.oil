strlen:    .function int 
           .param ptr str
           .local int len
           len = 0  
.wh0:      $t0:i = str[len * :i]  
           goto .od0 if $t0:i == '\0'
.do0:      $t1:i = len + 1
           len = $t1:i  
           goto .wh0  
.od0
           return len
           return
           .end
strcat:    .function 
           .param ptr dest
           .param ptr src
           .local int pos
           .local int srcix
           $t0:i = call strlen(dest)  
           pos = $t0:i  
           srcix = 0  
.wh1:      $t1:i = src[srcix * :i]  
           goto .od1 if $t1:i == '\0'
.do1:      $t2:i = src[srcix * :i]  
           dest[pos * :i] = $t2:i  
           $t3:i = pos + 1
           pos = $t3:i  
           $t4:i = srcix + 1
           srcix = $t4:i  
           goto .wh1  
.od1
           dest[pos * :i] = '\0'  
           return
           .end
main:      .function int 
           .param int argc
           .param ptr argv
           .local int length
           .local int argi
           .local ptr concat
           length = 0  
           argi = 1  
.wh2:      goto .od2 if argi > argc
.do2:      $t3:c = argv[argi * :c]  
           $t2:i = call strlen($t3:c)  
           $t1:i = length + $t2:i
           $t0:i = $t1:i + 2
           length = $t0:i  
           $t4:i = argi + 1
           argi = $t4:i  
           goto .wh2  
.od2
           call putstr("concat length = ")  
           call putint(length)  
           call putchr('\n')  
           concat = call malloc(length)  
           argi = 1  
           call putchr('[')  
.wh3:      goto .od3 if argi > argc
.do3:      call strcat(concat,"(")  
           $t5:c = argv[argi * :c]  
           call strcat(concat,$t5:c)  
           call strcat(concat,")")  
           $t6:i = argi + 1
           argi = $t6:i  
           goto .wh3  
.od3
           call putstr(concat)  
           call putstr("]\n")  
           return 0
           return
           .end
