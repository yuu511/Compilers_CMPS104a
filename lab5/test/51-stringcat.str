string_set[   0]:    7296690267450279874 0x5577c82b5cd8->"chr"
                   13348923631108115385 0x5577c82b6698->"putstr"
string_set[   5]:   12679825168387489523 0x5577c82b61f8->"num"
string_set[   6]:   16299098330955776099 0x5577c82bd248->""(""
                    1363220277528623646 0x5577c82b7478->"strlen"
string_set[   7]:    6814028903552918164 0x5577c82baa68->"array"
                    2297668033614959926 0x5577c82b7a98->"0"
string_set[   9]:   11100957978370531577 0x5577c82b8d18->"src"
string_set[  10]:    1583576014053361350 0x5577c82b6858->"str"
                   13497578857207247412 0x5577c82b5278->"string"
                   11597697714117577063 0x5577c82b79d8->"="
string_set[  11]:    1396094175042499165 0x5577c82bab28->"<"
string_set[  13]:   12868445110721718657 0x5577c82b58f8->")"
string_set[  16]:    5905351119970222309 0x5577c82b8698->"strcat"
                   11911796782241914524 0x5577c82b3fe8->"void"
                   10159970873491820195 0x5577c82b8358->"1"
string_set[  18]:   13581035385482130682 0x5577c82b7d58->"["
string_set[  19]:    5344662657899890615 0x5577c82b59b8->";"
string_set[  20]:   11868120863286193964 0x5577c82ba6e8->"main"
                   15816506932648033945 0x5577c82b51b8->"("
string_set[  21]:    3505610083221236292 0x5577c82baec8->"length"
                    2617056156435215158 0x5577c82b7ef8->"!="
string_set[  23]:    5266873771696973564 0x5577c82bb158->"argi"
string_set[  25]:    2178305242944497711 0x5577c82b9348->"srcix"
                   11166701219533395393 0x5577c82b6e78->"getstr"
string_set[  28]:   15761577654341137060 0x5577c82bdd08->""]\n""
string_set[  29]:    1370730858159036685 0x5577c82b8298->"+"
                   14635384904416252381 0x5577c82b8478->"return"
string_set[  34]:    8290819294854170237 0x5577c82bc698->"alloc"
                    7211130878751977441 0x5577c82b5b18->"putchr"
string_set[  36]:   11752130776608228127 0x5577c82b7fb8->"'\0'"
string_set[  37]:    6192799629375248351 0x5577c82bc3f8->"'\n'"
                   12291392672316406109 0x5577c82b5818->"line"
string_set[  39]:   15410491504089378546 0x5577c82bc078->""concat length = ""
                   18413295556186398326 0x5577c82b6038->"putint"
string_set[  40]:   10319632053171904911 0x5577c82b7c18->"while"
string_set[  44]:      21606623943142546 0x5577c82b77d8->"{"
string_set[  45]:    6142509188972423790 0x5577c82b53f8->""
string_set[  47]:   16541611849048157824 0x5577c82b8f98->"pos"
                   13686721134187162920 0x5577c82b84e8->"}"
string_set[  51]:   15141261116119706963 0x5577c82bc558->"concat"
                    3701972582333163920 0x5577c82ba8a8->"argc"
string_set[  53]:   12051031859636400188 0x5577c82b7e98->"]"
string_set[  57]:   17897127611569151174 0x5577c82bd828->"")""
                    9386170080689500517 0x5577c82b5338->"expr"
string_set[  58]:    4017852018407011111 0x5577c82b55d8->"file"
                    3729804957429652673 0x5577c82aecc8->"<<ROOT>>"
string_set[  59]:   16579916199511856337 0x5577c82b7178->"getln"
string_set[  60]:    9423661692722584567 0x5577c82b8ad8->"dest"
string_set[  61]:    4551451650890805270 0x5577c82bbc48->"2"
                   14643508047410943861 0x5577c82bacc8->"argv"
string_set[  63]:    9611428142580219387 0x5577c82b50b8->"fail"
string_set[  64]:   12597418120160561823 0x5577c82b5758->"int"
string_set[  70]:    9201800968376702500 0x5577c82b6b78->"getchr"
string_set[  71]:   14526079891305879776 0x5577c82bac68->">"
string_set[  72]:   13310716766371036271 0x5577c82b7918->"len"
string_set[  75]:    3012229484421721799 0x5577c82bcb18->"'['"
string_set[  78]:    4706206694200543916 0x5577c82b50f8->","
load_factor = 0.734
bucket_count = 79
max_bucket_size = 3
