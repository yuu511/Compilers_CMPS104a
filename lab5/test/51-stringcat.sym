strcat (7.14.1) {8} void function 
   dest (7.14.14) {9} string variable param lval 0
   src (7.14.27) {9} string variable param lval 1
   pos (7.15.4) {9} int variable local lval 2
   srcix (7.16.4) {9} int variable local lval 3

strlen (7.8.1) {7} int function 
   str (7.8.13) {8} string variable param lval 0
   len (7.9.4) {8} int variable local lval 1

getln (6.30.1) {6} string function 

putchr (6.24.1) {1} void function 

putstr (6.26.1) {3} void function 

fail (6.22.1) {0} void function 

getstr (6.29.1) {5} string function 

putint (6.25.1) {2} void function 

main (7.25.1) {9} int function 
   argc (7.25.11) {10} int variable param lval 0
   argv (7.25.21) {10} string array variable param lval 1
   length (7.26.4) {10} int variable local lval 2
   argi (7.27.4) {10} int variable local lval 3
   concat (7.35.4) {10} string variable local lval 4

getchr (6.28.1) {4} int function 

