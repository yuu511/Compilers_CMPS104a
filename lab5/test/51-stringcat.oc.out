CONDITIONAL WHILE STATEMENT SYMBOL TOK_NE 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_NE 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
